src/appleseed/renderer/kernel/rendering/pixelrendererbase.h

```diff
@@ -74,18 +78,48 @@ class PixelRendererBase
 
   protected:
     void on_pixel_begin(
-        const foundation::Vector2i& pi,
-        AOVAccumulatorContainer&    aov_accumulators);
+        const foundation::Vector2i&         pi,
+        const foundation::Vector2i&         pt,
+        const foundation::AABB2i&           tile_bbox,
+        AOVAccumulatorContainer&            aov_accumulators);
 
     void on_pixel_end(
-        const foundation::Vector2i& pi,
-        AOVAccumulatorContainer&    aov_accumulators);
+        const foundation::Vector2i&         pi,
+        const foundation::Vector2i&         pt,
+        const foundation::AABB2i&           tile_bbox,
+        AOVAccumulatorContainer&            aov_accumulators);
 
     void signal_invalid_sample();
 
   private:
-    size_t                          m_invalid_sample_count;
-    size_t                          m_invalid_pixel_count;
+    struct Parameters
+    {
+        const bool                  m_diagnostics;
+
+        explicit Parameters(const ParamArray& params)
+            : m_diagnostics(params.get_optional<bool>("enable_diagnostics", false))
+        {
+        }
+    };
+
+    size_t                                  m_invalid_sample_count;
+    size_t                                  m_invalid_pixel_count;
+    size_t                                  m_invalid_sample_aov_index;
+    std::unique_ptr<foundation::Tile>       m_invalid_sample_diagnostic;
+
+  protected:
+    const Parameters                m_base_params;
+};
+
```

Comment:
Add a second blank line before block comments. That's the only case where we use two blank lines.